Support knx operation types

marvin-w opened the pull request:
## Description:

Updated xknx version to 0.9.3
Adjusted climate component due to changes in the underlying library.

**Release notes**

- **Breaking change:**
  Refactored climate component to be able to support operation types. See the documentation for the new operation modes.
- the lovelace climate card can now be used with a knx climate device
- fixed a bug where updating the operation mode wouldn't update the lovelace climate card

**Related issue (if applicable):** fixes #14531

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7937


## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Commit:
Updated version to 0.9.3

Adjusted climate component due to changes in the underlying library.

homeassistant commented:
Hi @marvin-w, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#19546).

Once you do that we will be able to review and accept this pull request.

Thanks!

Julius2342 commented on homeassistant/components/climate/knx.py:
```diff
@@ -54,91 +61,120 @@
-        device = hass.data[DATA_KNX].xknx.devices[device_name]
-        entities.append(KNXClimate(device))
-    async_add_entities(entities)
+        device = hass.data[DATA_XKNX].xknx.devices[device_name]
```
DATA_XKNX -> DATA_KNX

Julius2342 commented on homeassistant/components/climate/knx.py:
```diff
@@ -5,15 +5,15 @@
+import homeassistant.helpers.config_validation as cv
 import voluptuous as vol
-
+from custom_components.xknx import ATTR_DISCOVER_DEVICES, DATA_XKNX
```
DATA_XKNX -> DATA_KNX

Julius2342 commented on homeassistant/components/climate/knx.py:
```diff
@@ -54,91 +61,120 @@
 
 @callback
-def async_add_entities_discovery(hass, discovery_info, async_add_entities):
+def async_add_devices_discovery(hass, discovery_info, async_add_devices):
```
I think this was changed unintentionally.

Julius2342 commented on homeassistant/components/climate/knx.py:
```diff
@@ -54,91 +61,120 @@
 
 @callback
-def async_add_entities_config(hass, config, async_add_entities):
+def async_add_devices_config(hass, config, async_add_devices):
```
I think this was changed unintentionally.

Commit:
Climate.KNX: fix updating view when operation mode is changed due to refactoring

MartinHjelmare reviewed (changes requested):
We can't accept this until the correct operation modes are used for the home assistant user interface, ie `current_operation`, `operation_list` and `async_set_operation_mode`.

marvin-w commented on homeassistant/components/climate/knx.py:
```diff
@@ -75,57 +90,82 @@ def async_add_entities_discovery(hass, discovery_info, async_add_entities):
 
 @callback
-def async_add_entities_config(hass, config, async_add_entities):
+def async_add_entities_config(hass, config, async_add_devices):
```
Reverted.

marvin-w commented on homeassistant/components/climate/knx.py:
```diff
@@ -195,20 +231,36 @@ def max_temp(self):
-        if self.device.supports_operation_mode:
-            return self.device.operation_mode.value
+        if self.device.mode.supports_operation_mode:
+            return self.device.mode.operation_mode.value
```
I've added a mapping to support this. Unfortunetly I don't think that we can support all available operation modes yet due to the fact that HA only supports a couple.

See also: http://www.sti.uniurb.it/romanell/Domotica_e_Edifici_Intelligenti/110504-Lez10a-KNX-Datapoint%20Types%20v1.5.00%20AS.pdf - Page 111

Commit:
Addressed review comments

marvin-w commented:
@MartinHjelmare: Thank you for reviewing this. I think I've addressed all review comments. One thing I'm not sure about is the mapping of the operation modes. I'm not particularly happy with some of the mappings since they don't fit the HA operation modes completly.

Also we can't support  all operation modes. I think this is ok for now and I'll take care of that at a later stage.

MartinHjelmare commented on homeassistant/components/climate/knx.py:
```diff
@@ -54,11 +64,33 @@
     vol.Optional(CONF_OPERATION_MODE_COMFORT_ADDRESS): cv.string,
+    vol.Optional(CONF_ON_OFF_ADDRESS): cv.string,
+    vol.Optional(CONF_ON_OFF_STATE_ADDRESS): cv.string,
+    vol.Optional(CONF_OPERATION_MODES): cv.ensure_list_csv,
```
Don't use ensure list csv. It's only for backwards compatibility. Use `cv.ensure_list`.

Make sure to validate the values of the list too.

marvin-w replied:
Done.

Commit:
Added validation for config.

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Please add a paragraph in the PR description about the breaking change for the release notes.

marvin-w commented:
@MartinHjelmare I added the paragraph in the PR description.

MartinHjelmare commented:
I tweaked the message a bit. Please make sure to mention in the docs how the mapping is done between home assistant operation mode and knx mode.
